src/screen/loading.rs

```diff
@@ -5,41 +5,38 @@ use bevy::prelude::*;
 
 use super::Screen;
 use crate::{
-    game::assets::{ImageAssets, SfxAssets, SoundtrackAssets},
+    game::assets::{AssetMap, ImageKey, SfxKey, SoundtrackKey},
     ui::prelude::*,
 };
 
 pub(super) fn plugin(app: &mut App) {
     app.add_systems(OnEnter(Screen::Loading), enter_loading);
-    app.add_systems(Update, check_all_loaded.run_if(in_state(Screen::Loading)));
+    app.add_systems(
+        Update,
+        continue_to_title.run_if(in_state(Screen::Loading).and_then(all_assets_loaded)),
+    );
 }
 
-fn enter_loading(mut commands: Commands, asset_server: Res<AssetServer>) {
+fn enter_loading(mut commands: Commands) {
     commands
         .ui_root()
         .insert(StateScoped(Screen::Loading))
         .with_children(|children| {
             children.label("Loading...");
         });
+}
 
-    // Preload assets so the game runs smoothly.
-    commands.insert_resource(ImageAssets::new(&asset_server));
-    commands.insert_resource(SfxAssets::new(&asset_server));
-    commands.insert_resource(SoundtrackAssets::new(&asset_server));
+fn all_assets_loaded(
+    asset_server: Res<AssetServer>,
+    image_map: Res<AssetMap<ImageKey>>,
+    sfx_map: Res<AssetMap<SfxKey>>,
+    soundtrack_map: Res<AssetMap<SoundtrackKey>>,
+) -> bool {
+    image_map.all_loaded(&asset_server)
+        && sfx_map.all_loaded(&asset_server)
+        && soundtrack_map.all_loaded(&asset_server)
```

Comment:
Would be cool to use polymorphism here to reduce duplication, but I don't think that's possible here, is it? Some thing like

```rust
[image_map as IdkWhichType, sfx_map, soundtrack_map].iter().all(...)
```

Reply:
I did think about this. I'm not sure there's a more elegant option than just calling `.all_loaded` on each one, without ballooning the complexity. If there is a way, it should be a separate PR probably.